webapp/src/utils/adapters/index.ts

```diff
@@ -27,3 +31,51 @@ export interface IMessageAdapter {
 export interface ITranslationAdapter {
   getTranslations(): Promise<TranslationMap>;
 }
+
+/** convert { [id]: { file, texts } } to { [id]: text } */
+export function dehydrateMessageMap(
+  translations: MessageMap,
+): Record<string, string> {
+  const output: Record<string, string> = {};
+  Object.entries(translations).forEach(([id, { text }]) => {
+    output[id] = text;
+  });
+  return output;
+}
+
+type TranslationFilenameMap = Record<
+  string, // lang
+  Record<
+    string, // filename
+    Record<
+      string, // msg Id
+      string // text
+    >
+  >
+>;
+
+/** convert TranslateMap to { [lang]: { [filename]: {[msgId]: text} } */
+export function groupByFilename(
+  translationMap: TranslationMap,
+): TranslationFilenameMap {
+  const output: TranslationFilenameMap = {};
+  Object.entries(translationMap).forEach(([lang, messageMap]) => {
+    Object.entries(messageMap).forEach(([msgId, { text, sourceFile }]) => {
+      if (!output[lang]) {
+        output[lang] = {};
+      }
+      if (!output[lang][sourceFile]) {
+        output[lang][sourceFile] = {};
+      }
+      const shortId = getShortId(msgId, sourceFile);
+      output[lang][sourceFile][shortId] = text;
+    });
+  });
+  return output;
+}
+
+function getShortId(msgId: string, sourceFile: string): string {
+  const sourceFileDotCount = sourceFile.replace('/', '.').split('.').length - 1;
+  const msgIdArray = msgId.split('.').splice(sourceFileDotCount);
+  return msgIdArray.join('.');
+}
```

Comment:
This function is complex enough that it should have a somewhat extensive test suite.